libia2/padding.ld

```diff
@@ -10,21 +10,24 @@
    __start_$SECTION and before __stop_$SECTION.
 */
 SECTIONS {
-    ia2_shared_rodata : {
-        . = ALIGN(4096);
+    ia2_shared_rodata ALIGN(4096): {
         __start_ia2_shared_rodata = .;
         *(ia2_shared_rodata)
-        . = ALIGN(4096);
+        /*
+           This weird alignment expression is required so that the section will
+           get elided if empty. Apparently ld won't drop a section that has a
+           non-zero aligned end, even if its size ends up being zero?
```

Comment:
My point from earlier is that `. = ALIGN(4096);` will not change the value of `.` (and presumably therefore not change the end of the section) if `.` is zero. I don't want to be a huge pain in the ass here, but I really don't understand what effect this `ALIGN` expression is having here and why.

My test:
test.ld:
```
SECTIONS
{
  . = 4096;
  "1_test_assign_4096" = .;
  . = ALIGN(4096);
  "2_test_align_4096_on_4096" = .;
  . = 0;
  "3_test_assign_0" = .;
  . = ALIGN(4096);
  "4_test_align_4096_on_0" = .;
  . = 1;
  "5_test_assign_1" = .;
  . = ALIGN(4096);
  "6_test_align_4096_on_1" = .;
  . = 0;
  . = ALIGN(1);
  "7_test_align_1_on_0" = .;
  . = 0;
  . = ALIGN(. != 0 ? 4096 : 1);
  "8_test_ternary_on_0" = .;
  . = 1;
  . = ALIGN(. != 0 ? 4096 : 1);
  "9_test_ternary_on_1" = .;

  . = 1-000;
  .text : { *(.text) }
  . = 0x8000000;
  .data : { *(.data) }
  .bss : { *(.bss) }
}
```
trivial.c:
```
int main(int argc, char** argv) {
	return 0;
}
```

harness: `gcc trivial.c -T test.ld -o test && nm test | fgrep test`
output:
```
0000000000001000 T 1_test_assign_4096
0000000000001000 T 2_test_align_4096_on_4096
0000000000000000 T 3_test_assign_0
0000000000000000 T 4_test_align_4096_on_0
0000000000000001 T 5_test_assign_1
0000000000001000 T 6_test_align_4096_on_1
0000000000000000 T 7_test_align_1_on_0
0000000000000000 T 8_test_ternary_on_0
0000000000001000 T 9_test_ternary_on_1
```

So if the effect on `.` of both `. = ALIGN(4096)` and `. = ALIGN(. != 0 ? 4096 : 1)` is identical, then what's going on here?

Reply:
The only way I got this to work was the combination of the first ALIGN on the section (not on . like your test) PLUS the weird ternary. I don't have a way to test handy but I can reduce this later.

Reply:
I tested the following

test.ld:

```SECTIONS
{
  . = 4096;
  "1_test_assign_4096" = .;
  . = ALIGN(4096);
  "2_test_align_4096_on_4096" = .;
  . = 0;
  "3_test_assign_0" = .;
  . = ALIGN(4096);
  "4_test_align_4096_on_0" = .;
  . = 1;
  "5_test_assign_1" = .;
  . = ALIGN(4096);
  "6_test_align_4096_on_1" = .;
  . = 0;
  . = ALIGN(1);
  "7_test_align_1_on_0" = .;
  . = 0;
  . = ALIGN(. != 0 ? 4096 : 1);
  "8_test_ternary_on_0" = .;
  . = 1;
  . = ALIGN(. != 0 ? 4096 : 1);
  "9_test_ternary_on_1" = .;
  .test10 ALIGN(4096): {
    . = 0;
    . = ALIGN(4096);
    "10_test_section_align_4096" = .;
  }
  test11 ALIGN(4096): {
    . = 0;
    . = ALIGN(. != 0 ? 4096 : 1);
    "11_test_section_ternary_on_0" = .;
  }
}
```

Harness: `gcc trivial.c -Wl,-T test.ld -o test && readelf -S test | grep test`

What we're doing is avoiding emission of the _section_, not the symbols.

Reply:
That does convince me that somehow, *despite* calculating the same value, the ternary does somehow allow the section to be omitted while the regular ALIGN doesn't. Neither replacing the ternary with `ALIGN(1)` nor with `ALIGN(4096)` lets the `.test11` section be omitted.

Some cases I tested just now:
- `. = ALIGN(. != 0 ? 4096 : 1);` allows section omission
- `. = ALIGN(. == 0 ? 1 : 4096);` does not
- `. = ALIGN(0 != . ? 4096 : 1);` does not
- `. = ALIGN(. ? 4096 : 1);` does not

Something Very Strange is happening here, and I don't like it. It's not just use of ternary, but some magic property of using that exact syntax within an `ALIGN`. It must be a feature since this is used in the default linker script, but it's not one I've ever heard of.

That said, you've demonstrated the necessity of this particular syntax and I'm fine with merging it as long as we have a comment that sufficiently exposes this arcanum.

Reply:
Out of curiosity, I tested with lld.

lld wants a `  . = 0;` before `.test10`, but otherwise does accept this linker script... however, it does not seem to obey the same undocumented, syntax-specific rules for omitting the section. The only way I can get it to omit the `.test11` section is by removing all the assignments in it.

Reply:
Spoooky. Well, I think we should still keep this for BFD, and we can investigate further.

Reply:
```suggestion
           This weird alignment expression is required so that the section will
           get elided if empty. ld.bfd special-cases this syntax to mean this,
           which is not documented but appears in its default linker script; see
           binutils commit 9d12f64cdc6540b75938097d7f4ab460bb346528
```

Reply:
Good sleuthing, I'll make that change when I get home, thanks.